network/netcore/src/transport/tcp.rs

```diff
@@ -109,6 +113,59 @@ impl Transport for TcpTransport {
     }
 }
 
+/// Try to lookup the dns name, then filter addrs according to the `IpFilter`.
+fn resolve_with_filter<'a>(
+    ip_filter: IpFilter,
+    dns_name: &'a str,
+    port: u16,
+) -> impl Future<Output = io::Result<impl Iterator<Item = SocketAddr> + 'a>> + 'a {
+    async move {
+        Ok(lookup_host((dns_name, port))
+            .await?
+            .filter(move |socketaddr| ip_filter.matches(socketaddr.ip())))
+    }
+}
+
+/// Note: we need to take ownership of this `NetworkAddress` (instead of just
+/// borrowing the `&[Protocol]` slice) so this future can be `Send + 'static`.
+async fn resolve_and_connect(addr: NetworkAddress) -> io::Result<TcpStream> {
+    let protos = addr.as_slice();
+
+    if let Some(((ipaddr, port), _addr_suffix)) = parse_ip_tcp(protos) {
+        // this is an /ip4 or /ip6 address, so we can just connect without any
+        // extra resolving or filtering.
+        TcpStream::connect((ipaddr, port)).await
+    } else if let Some(((ip_filter, dns_name, port), _addr_suffix)) = parse_dns_tcp(protos) {
+        // resolve dns name and filter
+        let socketaddr_iter = resolve_with_filter(ip_filter, dns_name.as_ref(), port).await?;
+        let mut last_err = None;
+
+        // try to connect until the first succeeds
+        for socketaddr in socketaddr_iter {
+            match TcpStream::connect(socketaddr).await {
+                Ok(stream) => return Ok(stream),
+                Err(err) => last_err = Some(err),
+            }
+        }
+
+        Err(last_err.unwrap_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "could not resolve to any address",
```

Comment:
Should we say what IP Filter we used for debugging purposes?

Reply:
good idea!